site/src/blog/typechat-programs.md

```diff
@@ -0,0 +1,302 @@
+
+This is the core idea of TypeChat Programs.
+All we need to do is specify a type set of functions that can be called, and TypeChat will guide a given language model to produce a small program represented in JSON.
+
+```ts
+import * as typechat from "typechat";
+// ...
+
+async function main() {
+    // ...
+
+    const model = typechat.createLanguageModel(process.env);
+    const schema = fs.readFileSync(path.join(__dirname, "someSchema.ts"), "utf8");
+
+    // Note we are using 'createProgramTranslator', not 'createJsonTranslator'.
+    const translator = typechat.createProgramTranslator(model, schema);
+    
+    const result = await translator.translate(
+        "Read 'foo.txt', trim its whitespace, and then save it to 'bar.txt'."
+    );
+    
+    const jsonProgram = typechat.getData(result);
+    console.log(JSON.stringify(jsonProgram, undefined, 2));
+}
+
+main();
+```
+
+This program outputs the following:
+
+```json
+{
+  "@steps": [
+    {
+      "@func": "readFile",
+      "@args": [
+        "foo.txt"
+      ]
+    },
+    {
+      "@func": "trimString",
+      "@args": [
+        {
+          "@ref": 0
+        }
+      ]
+    },
+    {
+      "@func": "writeFile",
+      "@args": [
+        "bar.txt",
+        {
+          "@ref": 1
+        }
+      ]
+    }
+  ]
+}
+```
+
+What's more, we can use the TypeScript compiler to further validate the response.
+TypeChat has the capability to transform this into a real TypeScript program, and we can even preview this by adding two lines to the example above:
+    
+```ts
+const moduleText = typechat.getData(typechat.createModuleTextFromProgram(jsonProgram));
+console.log(moduleText);
+```
+
+which will print out the following:
+
+```ts
+import { API } from "./schema";
+function program(api: API) {
+  const step1 = api.readFile("foo.txt");
+  const step2 = api.trimString(step1);
+  return api.writeFile("bar.txt", step2);
+}
+```
+
+During the program translation step, the TypeScript compiler is able to use this representation to ensure that *each of these steps is well-typed*.
+
+## Evaluating
+
+To make this useful, we still need to implement each of the `API` methods in some way.
+TypeChat has an `evaluateJsonProgram` function for this, and provides a callback with the method name and arguments for each step.
+
+```ts
+typechat.evaluateJsonProgram(jsonProgram, async (func, args) => {
+    switch (func) {
+        case "readFile": {
+            const [filePath] = args as [string];
+            return fs.readFileSync(filePath, { encoding: "utf8" });
+        }
+        case "writeFile": {
+            const [filePath, contents] = args as [string, string];
+            fs.writeFileSync(filePath, contents, { encoding: "utf8" });
+        }
+        case "trimString": {
+            const [input] = args as [string];
+            return input.trim();
+        }
+    }
+});
+```
+
```

Comment:
Here I would talk about the advantages of the callback model: The evaluator *never* calls any functions, that's all up to the callback. And the callback gets to do any amount of validation it deems necessary. In other words, the host is fully in control of what transpires.